lib/block_view.go

```diff
@@ -4726,135 +5590,780 @@ func (bav *UtxoView) _connectUpdateProfile(
+	}
 
-	} else {
-		// When there's no pre-existing profile entry we need to do more
-		// checks.
-		if len(txMeta.NewUsername) == 0 {
-			return 0, 0, nil, RuleErrorProfileUsernameTooShort
+	// Save a copy of the bid entry so that we can use it in the disconnect.
+	nftBidKey := MakeNFTBidKey(bidderPKID.PKID, txMeta.NFTPostHash, txMeta.SerialNumber)
+	prevNFTBidEntry := bav.GetNFTBidEntryForNFTBidKey(&nftBidKey)
+
+	if txMeta.SerialNumber != uint64(0) {
+		// Verify the NFT entry that is being bid on exists.
+		if nftEntry == nil || nftEntry.isDeleted {
+			return 0, 0, nil, RuleErrorNFTBidOnNonExistentNFTEntry
 		}
-		// We allow users to create profiles without a description or picture
-		// in the consensus code. If desired, frontends can filter out profiles
-		// that don't have these fields.
-		//
-		// Creator percentage and stake multiple are sufficiently checked above.
 
-		// In this case we need to set all the fields using what was passed
-		// into the transaction.
+		// Verify the NFT entry being bid on is for sale.
+		if !nftEntry.IsForSale {
+			return 0, 0, nil, RuleErrorNFTBidOnNFTThatIsNotForSale
+		}
 
-		// If below block height, user transaction public key.
-		// If above block height, use ProfilePublicKey if available.
-		profileEntryPublicKey := txn.PublicKey
-		if blockHeight > ParamUpdaterProfileUpdateFixBlockHeight {
-			profileEntryPublicKey = profilePublicKey
+		// Verify that the bidder is not the current owner of the NFT.
+		if reflect.DeepEqual(nftEntry.OwnerPKID, bidderPKID.PKID) {
+			return 0, 0, nil, RuleErrorNFTOwnerCannotBidOnOwnedNFT
 		}
 
-		newProfileEntry = ProfileEntry{
-			PublicKey:   profileEntryPublicKey,
-			Username:    txMeta.NewUsername,
-			Description: txMeta.NewDescription,
-			ProfilePic:  txMeta.NewProfilePic,
+		// Verify that the bid amount is greater than the min bid amount for this NFT.
+		// We allow BidAmountNanos to be 0 if there exists a previous bid entry. A value of 0 indicates that we should delete the entry.
+		if txMeta.BidAmountNanos < nftEntry.MinBidAmountNanos && !(txMeta.BidAmountNanos == 0 && prevNFTBidEntry != nil) {
+			return 0, 0, nil, RuleErrorNFTBidLessThanMinBidAmountNanos
+		}
+	}
 
-			CoinEntry: CoinEntry{
-				CreatorBasisPoints: txMeta.NewCreatorBasisPoints,
+	// Connect basic txn to get the total input and the total output without
+	// considering the transaction metadata.
+	totalInput, totalOutput, utxoOpsForTxn, err := bav._connectBasicTransfer(
+		txn, txHash, blockHeight, verifySignatures)
+	if err != nil {
+		return 0, 0, nil, errors.Wrapf(err, "_connectNFTBid: ")
+	}
 
-				// The other coin fields are automatically set to zero, which is an
-				// appropriate default value for all of them.
-			},
+	// Verify that the transaction creator has sufficient bitclout to create the bid.
+	spendableBalance, err := bav.GetSpendableBitcloutBalanceNanosForPublicKey(txn.PublicKey)
```

Comment:
I think we should keep this check  as it doesn't hurt, but it does seem kindof unnecessary. Someone can easily create the bid and then spend their balance, which makes it more about being helpful than about validation, the latter of which is what consensus is focused on.

Reply:
I'm not sure I agree.  This prevents someone from creating a ridiculous bid, like a 100M CLOUT bid.  That bid amount is obviously impossible to achieve based on the supply of bitclout (~10M) and this check prevents such a bid from making it past consensus.  We could add a "MaxBidAmount" but I like that this check also ensures that the person making the bid has access to the amount of clout that they claim, even if they do not leave it in the account.